packages/lib/gpcircuits/test/bounds.spec.ts

```diff
@@ -0,0 +1,268 @@
+import "mocha";
+import {
+  BoundsCheckModuleInputNamesType,
+  BoundsCheckModuleOutputNamesType
+} from "../src";
+import { circomkit } from "./common";
+
+function inInterval(input: bigint, bounds: bigint[]): boolean {
+  return bounds[0] <= input && input <= bounds[1];
+}
+
+// 'Critical intervals' are those intervals of field elements which are larger
+// than `nBits` bits yet still satisfy the constraints of the `InInterval`
+// circuit for given bounds [a,b].  These are the intervals [1 << nBits, 1 <<
+// nBits + a - 1] and [p - (1 << nBits) + b + 1, p-1], where p is the field
+// modulus. This procedure returns these as an array of closed intervals, where
+// empty intervals are omitted.
+function criticalIntervals(
+  bounds: [bigint, bigint],
+  nBits: number
+): [bigint, bigint][] {
+  if (
+    bounds.some((bound) => !inInterval(bound, [0n, (1n << BigInt(nBits)) - 1n]))
+  ) {
+    throw new TypeError("Invalid bounds");
+  }
+  const [a, b] = bounds;
+  return (
+    [
+      [1n << BigInt(nBits), (1n << BigInt(nBits)) + a - 1n],
+      [
+        21888242871839275222246405745257275088548364400416034343698204186575808495617n -
+          (1n << BigInt(nBits)) +
+          b +
+          1n,
+        21888242871839275222246405745257275088548364400416034343698204186575808495616n
+      ]
+    ] as [bigint, bigint][]
+  ).filter(([c, d]) => d - c >= 0);
+}
+
+// The complement interval of [0, 1 << n[ ∪ critical interval.
+function complementInterval(
+  bounds: [bigint, bigint],
+  nBits: number
+): [bigint, bigint] {
+  if (bounds.some((bound) => !inInterval(bound, [0n, 1n << BigInt(nBits)]))) {
+    throw new TypeError("Invalid bounds");
+  }
+  const [a, b] = bounds;
+  const c = (1n << BigInt(nBits)) + a;
+  const d =
+    21888242871839275222246405745257275088548364400416034343698204186575808495617n -
+    (1n << BigInt(nBits)) +
+    b;
+  return [c, d];
+}
+
+describe("bounds.BoundsCheckModule should work for intervals with 64-bit unsigned integer bounds", function () {
+  const sampleIntervals: [bigint, bigint][] = [
+    [0n, 0n],
+    [0n, 1n << 4n],
+    [1n << 5n, 1n << 29n],
+    [1n << 4n, 1n << 31n],
+    [0n, (1n << 64n) - 1n]
+  ];
+
+  const sampleValues = [0n, 1n << 5n, 1n << 16n, 1n << 30n, (1n << 64n) - 1n];
+
+  let circuit: WitnessTester<
+    BoundsCheckModuleInputNamesType,
+    BoundsCheckModuleOutputNamesType
+  >;
+
+  this.beforeAll(async () => {
+    circuit = await circomkit.WitnessTester("BoundsCheckModule", {
+      file: "bounds",
+      template: "BoundsCheckModule",
+      params: [64]
+    });
+  });
+
+  it("should yield expected output for 64-bit unsigned values and bounds", async () => {
+    for (const comparisonValue of sampleValues) {
+      for (const [minValue, maxValue] of sampleIntervals) {
+        await circuit.expectPass(
+          { comparisonValue, minValue, maxValue },
+          { out: BigInt(+inInterval(comparisonValue, [minValue, maxValue])) }
+        );
+      }
+    }
+  });
+
```

Comment:
Needs some testing of out-of-bounds values which fit within nBits, as opposed to the critical intervals which are based around larger values.  I suggest testing two ways:
- Some specific sample values (like the in-bounds examples above) which can test random values which aren't close to the bounds, aren't powers of 2, etc.
- An automated test which for each bound X, tests that X-1, X, and X+1 give the correct output value.  This will need some logic to skip -1 or 2^64 as test values, or to anticipate the correct results for them at least.